// @swisscom/sdx/dist/js/webcomponents/collection/components/flip/flip.js
export class Flip {
    constructor() {
        this.direction = "none";
    }
    componentWillLoad() {
        this.el.style.display = "inline-block";
        this.el.style.transform = `scaleX(${this.getFlipTransformStyle("x")}) scaleY(${this.getFlipTransformStyle("y")})`;
    }
    getFlipTransformStyle(axis) {
        const map = {
            none: {
                x: 1,
                y: 1
            },
            horizontal: {
                x: -1,
                y: 1
            },
            vertical: {
                x: 1,
                y: -1
            },
            both: {
                x: -1,
                y: -1
            }
        };
        return map[this.direction][axis];
    }
    render() {
        return h("slot", null);
    }
    static get is() { return "sdx-flip"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "direction": {
            "type": String,
            "attr": "direction"
        },
        "el": {
            "elementRef": true
        }
    }; }
}
