// sdxcode/dist/js/webcomponents/collection/components/button/button.js
export class Button {
    constructor() {
        this.theme = "primary";
        this.background = "light";
        this.disabled = false;
        this.href = "";
        this.label = "";
        this.iconName = "";
        this.iconSize = 1;
        this.srHint = "";
    }
    onTouchStart() {
    }
    getComponentClassNames() {
        return {
            button: this.theme !== "transparent",
            [this.theme]: true,
            colorless: this.background === "dark",
            disabled: this.disabled,
            invalid: this.valid === false
        };
    }
    getIconClassNames() {
        return {
            icon: true,
            [this.iconName]: true,
            [`icon--s${this.iconSize}`]: true
        };
    }
    getAriaExpanded() {
        if (this.ariaExpandedOnButton === true) {
            return "true";
        }
        else if (this.ariaExpandedOnButton === false) {
            return "false";
        }
        else {
            return undefined;
        }
    }
    hostData() {
        return {
            class: {
                transparent: this.theme === "transparent"
            }
        };
    }
    render() {
        return this.href
            ?
                h("a", { class: this.getComponentClassNames(), href: this.href, target: this.target },
                    this.iconName &&
                        h("i", { class: this.getIconClassNames(), "aria-hidden": "true" }),
                    this.label,
                    " ",
                    h("span", { class: "sr-only" }, this.srHint))
            :
                h("button", { class: this.getComponentClassNames(), disabled: this.disabled, "aria-expanded": this.getAriaExpanded() },
                    this.iconName &&
                        h("i", { class: this.getIconClassNames(), "aria-hidden": "true" }),
                    this.label,
                    " ",
                    h("span", { class: "sr-only" }, this.srHint));
    }
    static get is() { return "sdx-button"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "ariaExpandedOnButton": {
            "type": Boolean,
            "attr": "aria-expanded-on-button"
        },
        "background": {
            "type": String,
            "attr": "background"
        },
        "disabled": {
            "type": Boolean,
            "attr": "disabled"
        },
        "href": {
            "type": String,
            "attr": "href"
        },
        "iconName": {
            "type": String,
            "attr": "icon-name"
        },
        "iconSize": {
            "type": Number,
            "attr": "icon-size"
        },
        "label": {
            "type": String,
            "attr": "label"
        },
        "srHint": {
            "type": String,
            "attr": "sr-hint"
        },
        "target": {
            "type": String,
            "attr": "target"
        },
        "theme": {
            "type": String,
            "attr": "theme"
        },
        "valid": {
            "type": Boolean,
            "attr": "valid"
        }
    }; }
    static get listeners() { return [{
            "name": "touchstart",
            "method": "onTouchStart",
            "passive": true
        }]; }
    static get style() { return "/**style-placeholder:sdx-button:**/"; }
}
