// @swisscom/sdx/dist/js/webcomponents/collection/components/menu-flyout/menu-flyout-toggle/menu-flyout-toggle.js
import { MenuFlyoutActionTypes } from "../menu-flyout-store";
import { getStore, mapStateToProps } from "../../../core/helpers/store-helpers";
export class MenuFlyoutToggle {
    onClick() {
        this.toggle();
    }
    handleKeyDown(e) {
        const key = e.key;
        if (key === "Spacebar" || key === " " || key === "Enter") {
            e.preventDefault();
            this.toggle();
        }
    }
    componentWillLoad() {
        this.store = getStore(this);
        this.unsubscribe = mapStateToProps(this, this.store, [
            "display",
            "toggle"
        ]);
        this.dispatch({ type: MenuFlyoutActionTypes.setToggleEl, toggleEl: this.el });
    }
    componentDidUnload() {
        this.dispatch({ type: MenuFlyoutActionTypes.setToggleEl, toggleEl: undefined });
        this.unsubscribe();
    }
    dispatch(action) {
        if (this.store) {
            this.store.dispatch(action);
        }
    }
    getAriaExpanded() {
        return this.display === "open" ? "true" : "false";
    }
    render() {
        return (h("button", { class: "toggle", "aria-expanded": this.getAriaExpanded() },
            h("slot", null)));
    }
    static get is() { return "sdx-menu-flyout-toggle"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "display": {
            "state": true
        },
        "el": {
            "elementRef": true
        },
        "toggle": {
            "state": true
        }
    }; }
    static get listeners() { return [{
            "name": "click",
            "method": "onClick"
        }, {
            "name": "keydown",
            "method": "handleKeyDown"
        }]; }
    static get style() { return "/**style-placeholder:sdx-menu-flyout-toggle:**/"; }
}
