// @swisscom/sdx/dist/es6/charts/ChartFunctions.js
import * as tslib_1 from "tslib";
import DomElement from "../DomElement";
import { text } from "../DomFunctions";
var QUERY_DATA = ".js-data";
export function tryGetData(element) {
    var e_1, _a;
    var data = [];
    var elements = element.querySelectorAll(QUERY_DATA);
    try {
        for (var elements_1 = tslib_1.__values(elements), elements_1_1 = elements_1.next(); !elements_1_1.done; elements_1_1 = elements_1.next()) {
            var entry = elements_1_1.value;
            var value = parseFloat(entry.getAttribute("data-value"));
            var color = entry.getAttribute("data-color");
            var title = text(entry);
            var item = {
                title: title,
                value: value,
                color: color
            };
            data.push(item);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (elements_1_1 && !elements_1_1.done && (_a = elements_1.return)) _a.call(elements_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return data;
}
export function removeAllChildren(node) {
    while (node.firstChild) {
        node.removeChild(node.firstChild);
    }
}
export function createLegendItem(data) {
    var bullet = new DomElement("span")
        .addClass("bullet");
    if (isColor(data.color) === true) {
        bullet.setAttribute("style", "background-color: " + data.color + ";");
    }
    else {
        bullet.addClass(data.color);
    }
    var caption = new DomElement("span")
        .setHtml(data.title);
    return new DomElement("li")
        .appendChild(bullet)
        .appendChild(caption)
        .element;
}
export function isColor(str) {
    var pattern = /^#/i;
    return pattern.test(str);
}
