// sdxcode/dist/js/webcomponents/esm/es2017/build/4whqmc6f.sc.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

class ButtonGroup {
    constructor() {
        this.layout = "responsive";
    }
    getComponentClassNames() {
        return {
            component: true,
            [this.layout]: true
        };
    }
    hostData() {
        return {
            class: {
                [this.layout]: true
            }
        };
    }
    render() {
        return (h("div", { class: this.getComponentClassNames() },
            h("slot", null)));
    }
    static get is() { return "sdx-button-group"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "layout": {
            "type": String,
            "attr": "layout"
        }
    }; }
    static get style() { return "*.sc-sdx-button-group, .sc-sdx-button-group:after, .sc-sdx-button-group:before, .sc-sdx-button-group-h{-webkit-box-sizing:border-box;box-sizing:border-box}.sc-sdx-button-group-h   .component.sc-sdx-button-group{display:-ms-flexbox;display:flex;-ms-flex-wrap:wrap;flex-wrap:wrap;-ms-flex-direction:row-reverse;flex-direction:row-reverse;margin:-8px}.sc-sdx-button-group-h .sc-sdx-button-group-s > sdx-button{margin:8px}\@media (max-width:479px){.sc-sdx-button-group-h .sc-sdx-button-group-s > sdx-button{width:100%}}.sc-sdx-button-group-h   .fixed.sc-sdx-button-group, .sc-sdx-button-group-h   .fullwidth.sc-sdx-button-group{-ms-flex-direction:row;flex-direction:row}.sc-sdx-button-group-h   .fill.sc-sdx-button-group{-ms-flex-wrap:nowrap;flex-wrap:nowrap}.sc-sdx-button-group-h.fill .sc-sdx-button-group-s > sdx-button{width:100%;min-width:0;-ms-flex:0 1 auto;flex:0 1 auto}\@media (max-width:479px){.sc-sdx-button-group-h.fixed .sc-sdx-button-group-s > sdx-button{width:auto}}.sc-sdx-button-group-h.fullwidth .sc-sdx-button-group-s > sdx-button{width:100%}"; }
}

export { ButtonGroup as SdxButtonGroup };
