// sdxcode/dist/js/webcomponents/esm/es2017/build/4whqmc6f.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

class ButtonGroup {
    constructor() {
        this.layout = "responsive";
    }
    getComponentClassNames() {
        return {
            component: true,
            [this.layout]: true
        };
    }
    hostData() {
        return {
            class: {
                [this.layout]: true
            }
        };
    }
    render() {
        return (h("div", { class: this.getComponentClassNames() },
            h("slot", null)));
    }
    static get is() { return "sdx-button-group"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "layout": {
            "type": String,
            "attr": "layout"
        }
    }; }
    static get style() { return "*,:after,:before,:host{-webkit-box-sizing:border-box;box-sizing:border-box}:host .component{display:-ms-flexbox;display:flex;-ms-flex-wrap:wrap;flex-wrap:wrap;-ms-flex-direction:row-reverse;flex-direction:row-reverse;margin:-8px}:host ::slotted(sdx-button){margin:8px}\@media (max-width:479px){:host ::slotted(sdx-button){width:100%}}:host .fixed,:host .fullwidth{-ms-flex-direction:row;flex-direction:row}:host .fill{-ms-flex-wrap:nowrap;flex-wrap:nowrap}:host(.fill) ::slotted(sdx-button){width:100%;min-width:0;-ms-flex:0 1 auto;flex:0 1 auto}\@media (max-width:479px){:host(.fixed) ::slotted(sdx-button){width:auto}}:host(.fullwidth) ::slotted(sdx-button){width:100%}"; }
}

export { ButtonGroup as SdxButtonGroup };
