// sdxcode/dist/js/webcomponents/esm/es2017/build/wumytcyw.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

import { d as getSlot, e as installSlotObserver, f as isNativeSlot, g as getAllSlotChildNodes, h as parseFunction, i as closest, j as getPreviousFromList, k as getNextFromList } from './chunk-d1262b54.js';

class TextTruncate {
    doesSlotContainHTML() {
        return getAllSlotChildNodes(this.el).some((node) => node.nodeType === 1);
    }
    getComponentClassNames() {
        return {
            component: true,
            ellipsis: !this.doesSlotContainHTML()
        };
    }
    render() {
        return (h("div", { class: this.getComponentClassNames() },
            h("div", { class: "slot" },
                h("slot", null))));
    }
    static get is() { return "sdx-text-truncate"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "el": {
            "elementRef": true
        }
    }; }
    static get style() { return "*,:after,:before,:host{-webkit-box-sizing:border-box;box-sizing:border-box}:host{display:block;width:100%}.component.ellipsis .slot{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}"; }
}

export { TextTruncate as SdxTextTruncate };
