// sdxcode/dist/js/webcomponents/esm/es2017/build/ysqben9f.sc.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

import { b as MenuFlyoutActionTypes } from './chunk-bc34555f.js';
import { c as getStore, b as mapStateToProps } from './chunk-6a8011c5.js';

class MenuFlyoutToggle {
    onClick() {
        this.toggle();
    }
    handleKeyDown(e) {
        const key = e.key;
        if (key === "Spacebar" || key === " " || key === "Enter") {
            e.preventDefault();
            this.toggle();
        }
    }
    componentWillLoad() {
        this.store = getStore(this);
        this.unsubscribe = mapStateToProps(this, this.store, [
            "display",
            "toggle"
        ]);
        this.dispatch({ type: MenuFlyoutActionTypes.setToggleEl, toggleEl: this.el });
    }
    componentDidUnload() {
        this.dispatch({ type: MenuFlyoutActionTypes.setToggleEl, toggleEl: undefined });
        this.unsubscribe();
    }
    dispatch(action) {
        if (this.store) {
            this.store.dispatch(action);
        }
    }
    getAriaExpanded() {
        return this.display === "open" ? "true" : "false";
    }
    render() {
        return (h("button", { class: "toggle", "aria-expanded": this.getAriaExpanded() },
            h("slot", null)));
    }
    static get is() { return "sdx-menu-flyout-toggle"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "display": {
            "state": true
        },
        "el": {
            "elementRef": true
        },
        "toggle": {
            "state": true
        }
    }; }
    static get listeners() { return [{
            "name": "click",
            "method": "onClick"
        }, {
            "name": "keydown",
            "method": "handleKeyDown"
        }]; }
    static get style() { return "*.sc-sdx-menu-flyout-toggle, .sc-sdx-menu-flyout-toggle:after, .sc-sdx-menu-flyout-toggle:before, .sc-sdx-menu-flyout-toggle-h{-webkit-box-sizing:border-box;box-sizing:border-box}.toggle.sc-sdx-menu-flyout-toggle{background-color:transparent;color:#1781e3;border:none;font:inherit;outline:none;cursor:pointer}.toggle.sc-sdx-menu-flyout-toggle:focus, .toggle.sc-sdx-menu-flyout-toggle:hover{color:#0851da}"; }
}

export { MenuFlyoutToggle as SdxMenuFlyoutToggle };
