// sdxcode/dist/js/webcomponents/esm/es2017/build/zsqqvwws.sc.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

import { a as anime } from './chunk-16499f22.js';

class ScrollToTop {
    constructor() {
        this.animationDuration = 500;
        this.animationTranslationY = 112;
        this.position = "fixed";
        this.demo = false;
        this.visible = false;
    }
    componentWillLoad() {
        this.visible = this.isVisible();
        this.updateVisibility();
    }
    onWindowScroll() {
        this.visible = this.isVisible();
    }
    onTouchStart() {
    }
    isVisible() {
        return this.demo || this.hasScrolledOverWindowHeight();
    }
    hasScrolledOverWindowHeight() {
        return window.pageYOffset > window.innerHeight;
    }
    onClick() {
        this.scrollToTop();
    }
    onVisibleChanged() {
        this.updateVisibility();
    }
    updateVisibility() {
        if (this.visible) {
            anime({
                targets: this.el,
                duration: this.animationDuration,
                opacity: 1,
                translateY: 0,
                easing: "cubicBezier(0.550, 0.085, 0.320, 1)",
                begin: () => {
                    this.el.style.display = "block";
                    if (this.position === "fixed") {
                        this.el.style.position = "fixed";
                        this.el.style.bottom = "32px";
                        this.el.style.right = "32px";
                    }
                }
            });
        }
        else {
            anime({
                targets: this.el,
                duration: this.animationDuration,
                opacity: 0,
                translateY: this.animationTranslationY,
                easing: "cubicBezier(0.550, 0.085, 0.320, 1)",
                complete: () => {
                    this.el.style.display = "none";
                }
            });
        }
        this.el.forceUpdate();
    }
    scrollToTop() {
        window.scroll({ top: 0, behavior: "smooth" });
    }
    render() {
        return (h("button", { class: "scroll-to-top", "aria-hidden": "true" },
            h("sdx-icon", { "icon-name": "icon-arrow-up", size: 3 })));
    }
    static get is() { return "sdx-scroll-to-top"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "demo": {
            "type": Boolean,
            "attr": "demo"
        },
        "el": {
            "elementRef": true
        },
        "position": {
            "type": String,
            "attr": "position"
        },
        "visible": {
            "state": true,
            "watchCallbacks": ["onVisibleChanged"]
        }
    }; }
    static get listeners() { return [{
            "name": "window:scroll",
            "method": "onWindowScroll",
            "passive": true
        }, {
            "name": "touchstart",
            "method": "onTouchStart",
            "passive": true
        }, {
            "name": "click",
            "method": "onClick"
        }]; }
    static get style() { return "*.sc-sdx-scroll-to-top, .sc-sdx-scroll-to-top:after, .sc-sdx-scroll-to-top:before, .sc-sdx-scroll-to-top-h{-webkit-box-sizing:border-box;box-sizing:border-box}.sc-sdx-scroll-to-top-h{z-index:50000;display:none}.scroll-to-top.sc-sdx-scroll-to-top{border:2px solid #1781e3;border-radius:50%;background-color:hsla(0,0%,100%,.8);color:#1781e3;height:48px;width:48px;padding:6px;outline:none;cursor:pointer}.scroll-to-top.sc-sdx-scroll-to-top:hover{border-color:#0851da;color:#0851da}\@media (hover:none){.scroll-to-top.sc-sdx-scroll-to-top{border-color:#1781e3;color:#1781e3}}.scroll-to-top.sc-sdx-scroll-to-top:active, .scroll-to-top.sc-sdx-scroll-to-top:focus{border-color:#0851da;color:#0851da}.scroll-to-top.sc-sdx-scroll-to-top:active{-webkit-transform:scale(.98);transform:scale(.98)}"; }
}

export { ScrollToTop as SdxScrollToTop };
