// @swisscom/sdx/dist/js/webcomponents/collection/components/search/search.js
import * as wcHelpers from "../../core/helpers/webcomponent-helpers";
import { isDesktopOrLarger } from "../../core/helpers/breakpoint-helpers";
export class Search {
    constructor() {
        this.invokeSearchSubmitCallback = () => null;
        this.invokeValueChangeCallback = () => null;
        this.invokeChangeCallback = () => null;
        this.value = "";
        this.placeholder = "";
        this.srHint = "";
        this.srHintForButton = "";
    }
    searchSubmitCallbackChanged() {
        this.setInvokeSearchSubmitCallback();
    }
    valueChangeCallbackChanged() {
        this.setInvokeValueChangeCallback();
    }
    changeCallbackChanged() {
        this.setInvokeChangeCallback();
    }
    onWindowResizeThrottled() {
        if (this.resizeTimer) {
            clearTimeout(this.resizeTimer);
        }
        this.resizeTimer = setTimeout(() => {
            this.el.forceUpdate();
        }, 10);
    }
    componentWillLoad() {
        this.setInvokeSearchSubmitCallback();
        this.setInvokeValueChangeCallback();
    }
    submitSearch() {
        if (this.sdxInputEl) {
            this.invokeSearchSubmitCallback(this.sdxInputEl.value);
        }
    }
    onInputElChanged(value) {
        this.value = value;
        this.invokeValueChangeCallback(value);
        this.invokeChangeCallback(value);
    }
    setInvokeSearchSubmitCallback() {
        this.invokeSearchSubmitCallback = wcHelpers.parseFunction(this.searchSubmitCallback);
    }
    setInvokeValueChangeCallback() {
        this.invokeValueChangeCallback = wcHelpers.parseFunction(this.valueChangeCallback);
    }
    setInvokeChangeCallback() {
        this.invokeChangeCallback = wcHelpers.parseFunction(this.changeCallback);
    }
    isClearIconHidden() {
        return isDesktopOrLarger() || !this.value.length;
    }
    clear() {
        this.value = "";
    }
    render() {
        return (h("div", { class: "wrapper" },
            h("sdx-input", { value: this.value, srHint: this.srHint, type: "search", placeholder: this.placeholder, hitEnterCallback: () => this.submitSearch(), inputCallback: (value) => this.onInputElChanged(value), ref: (el) => this.sdxInputEl = el, role: "search", inputStyle: {
                    paddingRight: this.isClearIconHidden()
                        ? "48px"
                        : "88px"
                } }),
            h("div", { class: "search-button" },
                h("sdx-button", { theme: "transparent", onClick: () => this.submitSearch(), srHint: this.srHintForButton, iconName: "icon-search", iconSize: 3 })),
            h("div", { class: "clear-button" },
                h("sdx-animation", { animationName: this.isClearIconHidden() ? "hide" : "show" },
                    h("sdx-button", { theme: "transparent", onClick: () => this.clear(), iconName: "icon-close", "aria-hidden": "true" })))));
    }
    static get is() { return "sdx-search"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "changeCallback": {
            "type": String,
            "attr": "change-callback",
            "watchCallbacks": ["changeCallbackChanged"]
        },
        "el": {
            "elementRef": true
        },
        "placeholder": {
            "type": String,
            "attr": "placeholder"
        },
        "searchSubmitCallback": {
            "type": String,
            "attr": "search-submit-callback",
            "watchCallbacks": ["searchSubmitCallbackChanged"]
        },
        "srHint": {
            "type": String,
            "attr": "sr-hint"
        },
        "srHintForButton": {
            "type": String,
            "attr": "sr-hint-for-button"
        },
        "value": {
            "type": String,
            "attr": "value",
            "mutable": true
        },
        "valueChangeCallback": {
            "type": String,
            "attr": "value-change-callback",
            "watchCallbacks": ["valueChangeCallbackChanged"]
        }
    }; }
    static get listeners() { return [{
            "name": "window:resize",
            "method": "onWindowResizeThrottled",
            "passive": true
        }]; }
    static get style() { return "/**style-placeholder:sdx-search:**/"; }
}
