// @swisscom/sdx/dist/js/webcomponents/esm/es2017/build/gjvwliro.sc.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

class ValidationMessage {
    constructor() {
        this.valid = false;
        this.validationMessage = "";
    }
    getComponentClassNames() {
        return {
            component: true
        };
    }
    render() {
        return (h("div", { class: this.getComponentClassNames(), "aria-hidden": "true" },
            h("sdx-icon", { "icon-name": "icon-exclamation-mark" }),
            " ",
            this.validationMessage));
    }
    static get is() { return "sdx-validation-message"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "valid": {
            "type": Boolean,
            "attr": "valid"
        },
        "validationMessage": {
            "type": String,
            "attr": "validation-message"
        }
    }; }
    static get style() { return "*.sc-sdx-validation-message, .sc-sdx-validation-message:after, .sc-sdx-validation-message:before, .sc-sdx-validation-message-h{-webkit-box-sizing:border-box;box-sizing:border-box}.component.sc-sdx-validation-message{color:#d12;font-size:16px;margin-top:6px}.component.sc-sdx-validation-message   sdx-icon.sc-sdx-validation-message{-webkit-transform:scale(1.5);transform:scale(1.5)}"; }
}

export { ValidationMessage as SdxValidationMessage };
