// @swisscom/sdx/dist/js/webcomponents/esm/es2017/build/is1zereh.sc.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

import { d as getSlot, e as installSlotObserver, f as isNativeSlot, g as getAllSlotChildNodes, h as parseFunction, i as closest, j as getPreviousFromList, k as getNextFromList } from './chunk-d1262b54.js';
import { a as createAndInstallStore, b as mapStateToProps } from './chunk-6a8011c5.js';
import { a as InputGroupActionTypes, b as inputGroupReducer } from './chunk-45c4a97b.js';

class InputGroup {
    constructor() {
        this.invokeChangeCallback = () => null;
        this.componentDidLoadComplete = false;
        this.type = "radio";
        this.name = "";
        this.inline = false;
        this.label = "";
        this.required = false;
    }
    typeChanged() {
        this.store.dispatch({ type: InputGroupActionTypes.setTypeState, typeState: this.type });
    }
    changeCallbackChanged() {
        this.setInvokeChangeCallback();
    }
    nameChanged() {
        this.dispatchNameAction();
    }
    inlineChanged() {
        this.store.dispatch({ type: InputGroupActionTypes.setInline, inline: this.inline });
    }
    selectedInputItemElsChanged() {
        if (this.componentDidLoadComplete) {
            this.invokeChangeCallback(this.getSelection());
        }
    }
    selectNextInputItemElFromChanged() {
        if (this.selectNextInputItemElFrom) {
            let nextEl = getNextFromList(this.inputItemElsSorted, this.selectNextInputItemElFrom);
            while (nextEl !== this.selectNextInputItemElFrom && nextEl.disabled) {
                nextEl = getNextFromList(this.inputItemElsSorted, nextEl);
            }
            this.store.dispatch({
                type: InputGroupActionTypes.selectInputItemEl,
                inputItemEl: nextEl
            });
        }
    }
    selectPreviousInputItemElFromChanged() {
        if (this.selectPreviousInputItemElFrom) {
            let prevEl = getPreviousFromList(this.inputItemElsSorted, this.selectPreviousInputItemElFrom);
            while (prevEl !== this.selectPreviousInputItemElFrom && prevEl.disabled) {
                prevEl = getPreviousFromList(this.inputItemElsSorted, prevEl);
            }
            this.store.dispatch({ type: InputGroupActionTypes.selectInputItemEl, inputItemEl: prevEl });
        }
    }
    getSelection() {
        return this.selectedInputItemEls.map((inputItemEl) => inputItemEl.value);
    }
    componentWillLoad() {
        this.setInvokeChangeCallback();
        this.store = createAndInstallStore(this, inputGroupReducer, this.getInitialState());
        this.unsubscribe = mapStateToProps(this, this.store, [
            "typeState",
            "selectedInputItemEls",
            "selectNextInputItemElFrom",
            "selectPreviousInputItemElFrom",
            "inputItemElsSorted"
        ]);
        this.store.dispatch({ type: InputGroupActionTypes.setTypeState, typeState: this.type });
    }
    componentDidLoad() {
        if (this.name) {
            this.dispatchNameAction();
        }
        this.store.dispatch({ type: InputGroupActionTypes.setInline, inline: this.inline });
        this.componentDidLoadComplete = true;
    }
    componentDidUnload() {
        this.unsubscribe();
    }
    getInitialState() {
        return {
            typeState: "radio",
            nameState: "",
            inline: false,
            selectedInputItemEls: [],
            selectNextInputItemElFrom: undefined,
            selectPreviousInputItemElFrom: undefined,
            inputItemElsSorted: []
        };
    }
    dispatchNameAction() {
        this.store.dispatch({
            type: InputGroupActionTypes.setNameState,
            nameState: this.name
        });
    }
    setInvokeChangeCallback() {
        this.invokeChangeCallback = parseFunction(this.changeCallback);
    }
    hostData() {
        return {
            role: this.typeState === "radio" ? "radiogroup" : null
        };
    }
    render() {
        return (h("div", { class: "wrapper" },
            this.label && h("label", null,
                this.label,
                this.required ? h("span", { "aria-hidden": "true" }, " *") : null),
            h("slot", null),
            this.validationMessage &&
                h("sdx-validation-message", { validationMessage: this.validationMessage })));
    }
    static get is() { return "sdx-input-group"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "changeCallback": {
            "type": String,
            "attr": "change-callback",
            "watchCallbacks": ["changeCallbackChanged"]
        },
        "el": {
            "elementRef": true
        },
        "getSelection": {
            "method": true
        },
        "inline": {
            "type": Boolean,
            "attr": "inline",
            "watchCallbacks": ["inlineChanged"]
        },
        "inputItemElsSorted": {
            "state": true
        },
        "label": {
            "type": String,
            "attr": "label"
        },
        "name": {
            "type": String,
            "attr": "name",
            "watchCallbacks": ["nameChanged"]
        },
        "nameState": {
            "state": true
        },
        "required": {
            "type": Boolean,
            "attr": "required"
        },
        "selectedInputItemEls": {
            "state": true,
            "watchCallbacks": ["selectedInputItemElsChanged"]
        },
        "selectNextInputItemElFrom": {
            "state": true,
            "watchCallbacks": ["selectNextInputItemElFromChanged"]
        },
        "selectPreviousInputItemElFrom": {
            "state": true,
            "watchCallbacks": ["selectPreviousInputItemElFromChanged"]
        },
        "type": {
            "type": String,
            "attr": "type",
            "watchCallbacks": ["typeChanged"]
        },
        "typeState": {
            "state": true
        },
        "valid": {
            "type": Boolean,
            "attr": "valid"
        },
        "validationMessage": {
            "type": String,
            "attr": "validation-message"
        }
    }; }
    static get style() { return "*.sc-sdx-input-group, .sc-sdx-input-group:after, .sc-sdx-input-group:before, .sc-sdx-input-group-h{-webkit-box-sizing:border-box;box-sizing:border-box}label.sc-sdx-input-group{display:block;margin-bottom:4px;cursor:text;color:#666;font-size:16px}"; }
}

export { InputGroup as SdxInputGroup };
