// @swisscom/sdx/dist/js/webcomponents/esm/es2017/build/wraequsi.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

import { b as MenuFlyoutActionTypes } from './chunk-bc34555f.js';
import { b as mapStateToProps, c as getStore } from './chunk-6a8011c5.js';

class MenuFlyoutListItem {
    constructor() {
        this.selectable = true;
        this.href = "javascript:void(0);";
        this.disabled = false;
    }
    onClick(e) {
        if (this.disabled) {
            e.stopPropagation();
        }
    }
    componentWillLoad() {
        this.store = getStore(this);
        this.unsubscribe = mapStateToProps(this, this.store, [
            "directionState"
        ]);
    }
    componentDidLoad() {
        this.dispatch({ type: MenuFlyoutActionTypes.toggleArrowEl, arrowEl: this.arrowEl });
    }
    componentDidUnload() {
        this.dispatch({ type: MenuFlyoutActionTypes.toggleArrowEl, arrowEl: this.arrowEl });
        this.unsubscribe();
    }
    dispatch(action) {
        if (this.store) {
            this.store.dispatch(action);
        }
    }
    hostData() {
        return {
            class: {
                selectable: this.selectable && !this.disabled,
                disabled: this.disabled,
                [this.directionState]: true
            }
        };
    }
    getLinkClassNames() {
        return {
            body: true,
            disabled: this.disabled
        };
    }
    render() {
        return (h("div", { class: "item" },
            h("div", { class: "arrow", ref: (el) => this.arrowEl = el }),
            h("a", { href: this.href, class: this.getLinkClassNames() },
                h("slot", null))));
    }
    static get is() { return "sdx-menu-flyout-list-item"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "directionState": {
            "state": true
        },
        "disabled": {
            "type": Boolean,
            "attr": "disabled"
        },
        "el": {
            "elementRef": true
        },
        "href": {
            "type": String,
            "attr": "href"
        },
        "selectable": {
            "type": Boolean,
            "attr": "selectable"
        }
    }; }
    static get listeners() { return [{
            "name": "click",
            "method": "onClick",
            "capture": true
        }]; }
    static get style() { return "*,:after,:before,:host{-webkit-box-sizing:border-box;box-sizing:border-box}:host>.item>.arrow{display:none;position:absolute;background-color:#fff;width:14px;height:14px;-webkit-transform:rotate(45deg);transform:rotate(45deg)}:host{display:block}:host>.item>.body{position:relative;background-color:#fff;color:#1781e3;display:block;padding:12px 24px;text-align:center;text-decoration:none;-webkit-transition:border-bottom .2s cubic-bezier(.4,0,.6,1),color .2s cubic-bezier(.4,0,.6,1);transition:border-bottom .2s cubic-bezier(.4,0,.6,1),color .2s cubic-bezier(.4,0,.6,1)}:host(.selectable)>.item>.body{cursor:pointer}:host(:not(.selectable))>.item>.body{cursor:not-allowed}:host(.selectable:hover)>.item>.arrow,:host(.selectable:hover)>.item>.body{color:#fff;background-color:#1781e3!important}:host(.disabled)>.item>.body{color:#d6d6d6}:host(.disabled)>.item>.body.disabled,:host(.disabled)>.item>.body:disabled{cursor:not-allowed}:host(.disabled)>.item>.body.disabled:active,:host(.disabled)>.item>.body:disabled:active{pointer-events:none;-webkit-transform:none;transform:none}:host(:not(:last-of-type))>.item>.body{border-bottom:1px solid #e4e9ec}:host(.bottom-left:first-of-type)>.item>.arrow,:host(.bottom-right:first-of-type)>.item>.arrow{display:block;top:-7px;-webkit-box-shadow:-1px -1px 2px 0 rgba(0,0,0,.15);box-shadow:-1px -1px 2px 0 rgba(0,0,0,.15)}:host(.top-left:last-of-type)>.item>.arrow,:host(.top-right:last-of-type)>.item>.arrow{display:block;bottom:-7px;-webkit-box-shadow:1px 1px 2px 0 rgba(0,0,0,.15);box-shadow:1px 1px 2px 0 rgba(0,0,0,.15)}"; }
}

export { MenuFlyoutListItem as SdxMenuFlyoutListItem };
