// @swisscom/sdx/dist/js/webcomponents/esm/es2017/build/zuak0c3g.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../webcomponents.core.js';

import { d as getSlot, e as installSlotObserver, f as isNativeSlot, g as getAllSlotChildNodes, h as parseFunction, i as closest, j as getPreviousFromList, k as getNextFromList, l as isDesktopOrLarger } from './chunk-d1262b54.js';

class Search {
    constructor() {
        this.invokeSearchSubmitCallback = () => null;
        this.invokeValueChangeCallback = () => null;
        this.invokeChangeCallback = () => null;
        this.value = "";
        this.placeholder = "";
        this.srHint = "";
        this.srHintForButton = "";
    }
    searchSubmitCallbackChanged() {
        this.setInvokeSearchSubmitCallback();
    }
    valueChangeCallbackChanged() {
        this.setInvokeValueChangeCallback();
    }
    changeCallbackChanged() {
        this.setInvokeChangeCallback();
    }
    onWindowResizeThrottled() {
        if (this.resizeTimer) {
            clearTimeout(this.resizeTimer);
        }
        this.resizeTimer = setTimeout(() => {
            this.el.forceUpdate();
        }, 10);
    }
    componentWillLoad() {
        this.setInvokeSearchSubmitCallback();
        this.setInvokeValueChangeCallback();
    }
    submitSearch() {
        if (this.sdxInputEl) {
            this.invokeSearchSubmitCallback(this.sdxInputEl.value);
        }
    }
    onInputElChanged(value) {
        this.value = value;
        this.invokeValueChangeCallback(value);
        this.invokeChangeCallback(value);
    }
    setInvokeSearchSubmitCallback() {
        this.invokeSearchSubmitCallback = parseFunction(this.searchSubmitCallback);
    }
    setInvokeValueChangeCallback() {
        this.invokeValueChangeCallback = parseFunction(this.valueChangeCallback);
    }
    setInvokeChangeCallback() {
        this.invokeChangeCallback = parseFunction(this.changeCallback);
    }
    isClearIconHidden() {
        return isDesktopOrLarger() || !this.value.length;
    }
    clear() {
        this.value = "";
    }
    render() {
        return (h("div", { class: "wrapper" },
            h("sdx-input", { value: this.value, srHint: this.srHint, type: "search", placeholder: this.placeholder, hitEnterCallback: () => this.submitSearch(), inputCallback: (value) => this.onInputElChanged(value), ref: (el) => this.sdxInputEl = el, role: "search", inputStyle: {
                    paddingRight: this.isClearIconHidden()
                        ? "48px"
                        : "88px"
                } }),
            h("div", { class: "search-button" },
                h("sdx-button", { theme: "transparent", onClick: () => this.submitSearch(), srHint: this.srHintForButton, iconName: "icon-search", iconSize: 3 })),
            h("div", { class: "clear-button" },
                h("sdx-animation", { animationName: this.isClearIconHidden() ? "hide" : "show" },
                    h("sdx-button", { theme: "transparent", onClick: () => this.clear(), iconName: "icon-close", "aria-hidden": "true" })))));
    }
    static get is() { return "sdx-search"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "changeCallback": {
            "type": String,
            "attr": "change-callback",
            "watchCallbacks": ["changeCallbackChanged"]
        },
        "el": {
            "elementRef": true
        },
        "placeholder": {
            "type": String,
            "attr": "placeholder"
        },
        "searchSubmitCallback": {
            "type": String,
            "attr": "search-submit-callback",
            "watchCallbacks": ["searchSubmitCallbackChanged"]
        },
        "srHint": {
            "type": String,
            "attr": "sr-hint"
        },
        "srHintForButton": {
            "type": String,
            "attr": "sr-hint-for-button"
        },
        "value": {
            "type": String,
            "attr": "value",
            "mutable": true
        },
        "valueChangeCallback": {
            "type": String,
            "attr": "value-change-callback",
            "watchCallbacks": ["valueChangeCallbackChanged"]
        }
    }; }
    static get listeners() { return [{
            "name": "window:resize",
            "method": "onWindowResizeThrottled",
            "passive": true
        }]; }
    static get style() { return "\@charset \"UTF-8\";*,:after,:before,:host{-webkit-box-sizing:border-box;box-sizing:border-box}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0,0,0,0);border:0}.sr-only-focusable:active,.sr-only-focusable:focus{position:static;width:auto;height:auto;margin:0;overflow:visible;clip:auto}.wrapper{position:relative}.wrapper .clear-button,.wrapper .search-button{position:absolute;top:0}.wrapper .search-button{right:0;padding:7px}.wrapper .clear-button{right:40px;padding:11px 16px}"; }
}

export { Search as SdxSearch };
